### PR TITLE
🚀 Typescript 4 

### PR DESCRIPTION
**Pull request by nicolasgramlich**
# ToDo
 - [ ] Investigate TODOs in PR changes
 - [ ] Tests run green locally, but don't "exit". Probably a dangling promise or DB connection!?
    - image

### CODE REVIEW

File: src/classes/context.ts [hunk 1]
```diff
@@ -9,15 +9,15 @@ export class Context<T extends {} = any> {
   }
 
   public as<U extends T>(): Context<U> {
-    return this as Context<U>;
+    return this as unknown as Context<U>;
```
Comment (SylvainEstevez) [inline src/classes/context.ts:12]: Really necessary ? 🤔

File: src/classes/sequelize-service.ts [hunk 2]
```diff
@@ -1,76 +1,77 @@
...
 
   public async create<KC extends keyof C = keyof {}>(object: W, options: TCreateOptions<R, C, KC> = {}): Promise<R & Pick<C, KC>> {
     return await SequelizeService.try(async () => {
       return await this.transaction(options, async () => {
-        const session = <ISession<W, R, C>>new CreateSession<W, R, C, KC>([object], options, this);
+        const createSession = new CreateSession<W, R, C, KC>([object], options, this);
+        const session = createSession as ISession<W, R, C>;
 
-        await this.executeHook(Hook.WILL_CREATE, session, this._beforeCreate.bind(this));
+        await this.executeHook(Hook.WILL_CREATE, session, this._beforeCreate.bind(this, createSession)); // TODO Make sure the bind calls actually work...
```
Comment (SylvainEstevez) [inline src/classes/sequelize-service.ts:66]: Wow

**Review by SylvainEstevez — COMMENTED** [review]
Just a quick look let me see that you're still working on it :) Ideally we would get rid of the `as unknown as ...` types, but I guess they're now required :/

File: src/classes/context.ts (hunk 1 above)
Comment (nicolasgramlich) [reply src/classes/context.ts:12]: Yeah, there might be another way to constrain `U` somehow, but if you leave it out, the compiler will explicitly tell you to cast it to unknown first.

**Comment by nicolasgramlich** [comment]
@SylvainEstevez I might need some help on this one. You know this in and out.

**Comment by SylvainEstevez** [comment]
I'm really too busy these days to look at this though. Let's wait for after the VoC release end of October.

File: src/classes/context.ts (hunk 1 above)
Comment (SylvainEstevez) [reply src/classes/context.ts:12]: Yeah actually the whole types can now be REALLY simplified thanks to `infer`. Redoing them :)

**Pull request merged**